### ISSUE TITLE
dag_processing.processes should be a gauge, not a counter

### ISSUE THREAD

**Issue by Jeoffreybauvin** [apache/airflow#36188]
### What do you see as an issue?

I wasn't sure if it's a doc issue or a code issue.

For me, dag_processing.processes should be considered as a gauge, not a counter, due to negative values.

`statsd  | ts=2023-12-12T15:55:25.099Z caller=exporter.go:126 level=debug msg="counter must be non-negative value" metric=airflow_dag_processing_processes event_value=-1`

### Solving the problem

Define dag_processing.processes as a gauge.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
@Jeoffreybauvin Defining `dag_processing.processes` as a gauge is not simple because it's relative, so we use `increment` and `decrement` to update it, and these methods are not available for gauge.

Also, it's fine to have negative values in the counter, but it looks like the exception is raised in Airflow. I will check if there is a problem in the counter, or we just need to remove the validation check.

**Comment by hussein-awala** [comment]
I just checked OTEL code source and found that it doesn't allow negative results in all its counters (including `UpDownCounter`, which we use); maybe we need to implement a class to support relative Gauges with increase/decrease methods.
@ferruzzi WDYT?

**Comment by ferruzzi** [comment]
Sorry for the delay, I've been dealing with laptop issues and may be slow to reply.

From the [Airflow Metrics docs}(https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html), `dag_processing.processes` represents:

```Relative number of currently running DAG parsing processes (ie this delta is negative when, since the last metric was sent, processes have completed)```

In OTel, a Counter is an incremental value, like the odometer in your car, it can only go up.  A Gauge is like your speedometer, it only shows the current value.

Neither of these feel too accurate to the intent of the metric.  The [OTel specifications](https://opentelemetry.io/docs/specs/otel/metrics/api/#updowncounter), doesn't say an UpDownCounter must be positive.  If that is actually true in the python implementation, that may be a bug.

**Comment by dirrao** [comment]
I think we can publish the current number of running parsing processors as a gauge. I see it being updated if there is a new processor added or stopped/terminated.

**Comment by potiuk** [comment]
> I think we can publish the current number of running parsing processors as a gauge. I see it being updated if there is a new processor added or stopped/terminated.

SGTM

**Comment by Bowrna** [comment]
could i pick up this issue?

**Comment by dirrao** [comment]
> could i pick up this issue?

Yes. you can pick this task.

**Comment by Bowrna** [comment]
@dirrao @potiuk @ferruzzi @hussein-awala Do you think we have to fix in the existing key and edit the docs to or write this in new key and have a backward compatibility and slowly deprecate it in the upcoming release?

**Comment by hussein-awala** [comment]
Yes, a new metric will be better to avoid introducing a breaking change.

**Comment by Bowrna** [comment]
cool let me start this work by this weekend. its a tough past 3 weeks. hope to finish it soon
